### ISSUE TITLE
New annotation: psalm-self, as a pre-condition for the type of self when calling a method

### ISSUE THREAD

**Issue by olleharstedt** [vimeo/psalm#3740]
Use-case: A type-safe query builder:

```php
<?php
    
class Query
{   
    /**
     * @var string|null 
     */
    public $from;
        
    /**
     * @param string|null $from
     */
    public function __construct($from = null)
    {
        $this->from = $from;
    }
  
    /**
     * @param string $from
     * @return void
     * @psalm-self-out Query<string>
     */
    public function set($from)
    {
        $this->from = $from;
    }
  
    /**
     * @psalm-self Query<string>
     * @return string
     */
    public function execute()
    {
        return 'result';
    }
}

$q = new Query();
$q->execute();  // Should show error
$q->set('tablename');
$q->execute();  // OK
```

Another alternative would be using $this together with param: `@param Query<string> $this`.

Thoughts? Using `@psalm-self-out` to change interface of $this will not work, because there will be multiple fields in the query object (both "select" and "from" need to be set at least to be a valid query, and this cannot easily be controlled by switching interface of "self" back and forth).

Related discussion: https://github.com/vimeo/psalm/issues/3207#issuecomment-650061673

Can make a PR if accepted.

**Comment by muglug** [comment]
I think I prefer something more explicit – using a `@psalm-typestate` annotation, where a typestate is _similar_ to a template, with the key difference that it's mutable.

```php
<?php
   
/**
 * @psalm-typestate TTable as ?string
 */
class Query
{   
    /**
     * @var TTable
     */
    public $from;
        
    /**
     * @param TTable $from
     */
    public function __construct($from = null)
    {
        $this->from = $from;
    }
  
    /**
     * @template TNewTable as string
     * @param TNewTable $from
     * @return void
     * @psalm-self-out Query<TNewTable>
     */
    public function from($from)
    {
        $this->from = $from;
    }
  
    /**
     * @psalm-if-typestate TTable string
     * @return string
     */
    public function execute()
    {
        return 'result';
    }
}

$q = new Query();
$q->execute(); // fails
$q->from('tablename');
$q->execute(); // succeeds
```

**Comment by olleharstedt** [comment]
For reference, here's the same concept with templates: https://psalm.dev/r/5281d54b75 Pretty awkward. :)

> I think I prefer something more explicit – using a @psalm-typestate annotation, where a typestate is similar to a template, with the key difference that it's mutable.

Interesting. Just have to think if we block out any use-case with this restriction. :thinking:

**Comment by olleharstedt** [comment]
I like typestate as mutable templates.

One downside with `@psalm-if-typestate TTable string` is that it can't interact with templates. Even with immutable templates, maybe there's a use-case for different behaviour depending on if `T` is an int or a string? If I can think of one... :laughing: Maybe inheritance should be used in that case instead.

I agree that using `@psalm-self-out Query<string>` without defining a template (or typesate) for the class is kind of an abuse of syntax.

**Comment by muglug** [comment]
> One downside with `@psalm-if-typestate TTable string` is that it can't interact with templates.

What if generics are essentially a refinement of typestates? So all template param `TFoo` are also typestates, but the inverse is not true.

**Comment by muglug** [comment]
Also the thing I didn’t quite understand here is that typestate analysis can be applied to PHP code to make *existing* business logic errors into type errors.

**Comment by muglug** [comment]
Here's a more fleshed-out idea of a query builder. I think I prefer `@psalm-this-out` to `@psalm-self-out`, as only `$this` changes, not the mapping of `self`. 

```php
/**
 * @psalm-typestate T as QueryType::*
 */
class QueryBuilder {
  /**
    * @psalm-this-out self<QueryType::HAS_FROM>
    */
  public function __construct(string $table) : static { ... }

  /**
    * @psalm-if-this-is self<QueryType::HAS_FROM>
    */
  public function join($table, $on) : static { ... }

  /**
    * @psalm-if-this-is self<QueryType::HAS_FROM>|self<QueryType::HAS_WHERE>
    * @psalm-this-out self<QueryType::HAS_WHERE>
    */
  public function where($condition) : static { ... }

  /**
    * @psalm-if-this-is self<QueryType::HAS_FROM>|self<QueryType::HAS_WHERE>|self<QueryType::HAS_ORDER>
    * @psalm-this-out self<QueryType::HAS_ORDER>
    */
  public function orderBy($column, $direction) : static { ... }

  /**
    * @psalm-if-this-is self<QueryType::HAS_FROM>|self<QueryType::HAS_WHERE>|self<QueryType::HAS_ORDER>
    * @psalm-this-out self<QueryType::HAS_LIMIT>
    */
  public function limit($start, $length) : static { ... }
}

class QueryType {
  const HAS_FROM= 0;
  const HAS_WHERE = 1;
  const HAS_ORDER = 2;
  const HAS_LIMIT = 2;
}
```

cc @Ocramius @marcosh who may have thoughts

**Comment by Ocramius** [comment]
I must say that the last example makes the most sense, while mutable query builders are mostly just mistakes waiting to happen.

I saw a good talk about typed query builders in kotlin, some years ago: lemme fish it out.

**Comment by olleharstedt** [comment]
NB: You don't need to include WHERE, ORDER and LIMIT in the typestate, only SELECT and FROM, since that's what required to make execute() not faulty (for a SELECT query - let's disregard INSERT and UPDATE, they probably need to be different classes).

**Comment by olleharstedt** [comment]
> What if generics are essentially a refinement of typestates? So all template params TFoo are also typestate params, but all typestate params are not necessarily template params

My thought was that if we use `@param Foo<TBar> $this` or `@psalm-this Foo<TBar>` we don't have to care about the relationship between typestates and templates, since they are written the same way? Or both have the same abstract super-class.

**Comment by Ocramius** [comment]
Check https://twitter.com/breandan/status/1279387001972801541 meanwhile 👍

**Comment by olleharstedt** [comment]
Not sure we're talking about the same thing. One of the links shows this:

```kotlin
fun getQueryResults(context: DSLContext): Result<*>? {
    return context
        .select(AUTHOR.FIRST_NAME, AUTHOR.LAST_NAME, BOOK.TITLE)
        .from(AUTHOR)
        .join(BOOK).on(BOOK.AUTHOR_ID.eq(AUTHOR.ID))
        .where(AUTHOR.LAST_NAME.contains("Iverson")).fetch()
}
```

But it's not clear if it would fail at compilation if `select` or `from` was left out.

https://github.com/breandan/type-safety/blob/master/src/main/kotlin/answers/Answers.kt

**Comment by marcosh** [comment]
In the case of immutable classes, if I get it right, `@psalm-this-out` could actually be replaced by `@psalm-return`, and `@psalm-if-this-is` would be equivalent to what I proposed in #3207, right?

I wonder if it makes sense to allow to use such a mechanism also for mutable classes or if it is better to promote it only for immutable ones. In the last case, the distinction between `templates` and `typestates` would become irrelevant

**Comment by olleharstedt** [comment]
@marcosh Well, the point of typestate is exactly to lift mutable state into the type system. :)

**Comment by olleharstedt** [comment]
Also note that if this pattern is to be applied to already existing code, it must not change the constructor of the class.

**Comment by olleharstedt** [comment]
Another example by me: https://psalm.dev/r/c18e5558a1

What's needed to make it idiomatic is, as discussed:

* Mutable templates (to remove `@psalm-suppress InvalidPropertyAssignmentValue`)
* Checking the type of `$this` before calling method execute() (to lift execute() into the class)

**Comment by marcosh** [comment]
> @marcosh Well, the point of typestate is exactly to lift mutable state into the type system. :)

I get that, I was just asking whether Psalm should do that or allow it only for immutable classes

**Comment by Ocramius** [comment]
Same question here: it feels like a really weird edge case, whereas reasoning around immutable objects turns out to be much simpler.

**Comment by olleharstedt** [comment]
> Same question here: it feels like a really weird edge case, whereas reasoning around immutable objects turns out to be much simpler.

How would you do a query builder with immutable objects?

**Comment by Ocramius** [comment]
The builder isn't the actual executable, you get something back from it (built thing) and then operate on top of it.

As for immutable builders, I do use them (and write them) for my own code all the time, often criteria API around a small search DSL. You can look at PSR-7 messages for examples of that: good/simple way to assemble an HTTP request/response 👍

I do understand that current builders out there are mostly mutable, the question is whether that should be endorsed/facilitated (IMO no)

**Comment by olleharstedt** [comment]
Can you give me a link?

Looking at messages (https://www.php-fig.org/psr/psr-7/):

```php
$request = $request
    ->withMethod('OPTIONS')
    ->withRequestTarget('*')
    ->withUri(new Uri('https://example.org/'));
```

By setting a `@param Request<HasMethod> $this` on `__toString` (or however the request is used), we could statically (during Psalm analysis) make sure that `withMethod()` is not forgotten before the request is used.

**Comment by Ocramius** [comment]
Absolutely, it is possible to improve types there, but as @marcosh says, by combining `@psalm-return` and `@param`, without any additional complexity in the interpretation of `$this`

**Comment by olleharstedt** [comment]
> Absolutely, it is possible to improve types there, but as @marcosh says, by combining `@psalm-return` and `@param`, without any additional complexity in the interpretation of `$this`

Well, can you show an example on how? :)

**Comment by Ocramius** [comment]
```php
/** @psalm-template Filter of FilterExpression */
interface MyDSL
{
    /** 
     * @psalm-param Filter $e
     * @psalm-return static<Filter>&self
     */
    function withFilter(FilterExpression $e) :self;
}

/** @psalm-param MyDSL<FilterWithAtLeast3Chars> $e */
function search(MyDSL $e): Result { ... }

**Comment by olleharstedt** [comment]
I'd argue that putting `search` outside of MyDSL is not idiomatic. Would you create a separate class with only the search() method?

**Comment by Ocramius** [comment]
Most certainly.


The DSL is pure (and marked `@psalm-immutable`), while the runtime operation is very much full of side effects.

Keeping those separated allows you to implement different adapters and strategies for search (cache, no cache, sql, etc.).

One is domain logic, the other is infrastructure code.

**Comment by olleharstedt** [comment]
It's a good argument. Problem is that no code looks like that in the wild. :thinking:

**Comment by Ocramius** [comment]
Yes, and we need to get better at it: hence why we should probably endorse an approach that leads to better long-term results.

**Comment by marcosh** [comment]
I'd say that the following would look quite natural to do to me:

```php
class Status
{
    const INITIAL_STATUS = 0;
    const FINAL_STATUS = 1;
}

/**
 * @template S of Status::*
 */
class Foo
{
    /**
     * @return self<Status::INITIAL_STATUS>
     */
    public function __construct()

    /**
     * @constraint S of Status::INITIAL_STATUS
     * @return self<Status::FINAL_STATUS>
     */
    public function evolve() { ... } // returns new instance

    /**
     * @constraint S of Status::FINAL_STATUS
     */
    public function use() { ... }
}

$foo = new Foo();
$foo->use() // breaks;

$foo = $foo->evolve();
$foo->use(); // works
```

basically for every method we would like to be able to say what goes out, and this is done using the standard `@return`, but we want to control also what comes in, not only as function parameters, but also as the hidden state of the class. To do this we need something which allows us to narrow the type of a constraint.

I'm definitely not attached to the `constraint` name, any choice is basically fine by me, but that is the name that resonates better for me (see https://en.wikibooks.org/wiki/Haskell/Classes_and_types#Type_constraints)

**Comment by olleharstedt** [comment]
> Yes, and we need to get better at it: hence why we should probably endorse an approach that leads to better long-term results.

Well, is that a mission statement for Psalm, to promote better idioms?

(Btw, mutability is only a problem when there's aliasing. I wrote a little about ownership here: https://github.com/vimeo/psalm/issues/3734#issuecomment-653235121)

**Comment by muglug** [comment]
> Well, is that a mission statement for Psalm, to promote better idioms?

So that’s an interesting thought. Psalm is no longer defined solely by my opinion of it, but I think Psalm (the tool) should allow you to write code using a number of commonly-used idioms, providing ways to turn business logic errors into type errors in all sorts of scenarios.

That’s part of the reason for annotations like `@psalm-template` and  `@psalm-assert`, which can be added to existing code without changing any other underlying behaviour.

This seems to fit that mould rather nicely, though it’s definitely much more of an edge-case than those others.

**Comment by olleharstedt** [comment]
OK, so maybe I make a PR and we discuss more there? As with `@psalm-self-out`, it's probably best if it's undocumented until we've decided what to call the tags.

**Comment by olleharstedt** [comment]
PR: https://github.com/vimeo/psalm/pull/3782

**Comment by olleharstedt** [comment]
This design pattern can also prevent you from starting a database transaction twice, but can't enforce that it's being committed.

**Comment by olleharstedt** [comment]
Another use-case is to freeze objects on the go to make them immutable (just FYI):

```php
<?php

class Frozen {}

class Unfrozen {}

/**
 * @template T of Frozen|Unfrozen
 */
class Foo
{
    /**
     * @var T
     */
    private $state;

    /**
     * @param T $state
     */
    public function __construct($state)
    {
        $this->state = $state;
    }

    /**
     * @param string $name
     * @param mixed $val
     * @psalm-if-this-is Foo<Unfrozen>
     * @return void
     */
    public function set($name, $val)
    {
        $this->$name = $val;
    }

    /**
     * @return Foo<Frozen>
     */
    public function freeze()
    {
        /** @var Foo<Frozen> */
        $f = clone $this;
        return $f;
    }
}

$f = new Foo(new Unfrozen());
$f->set('asd', 10);
$g = $f->freeze();
$g->set('asd', 20);  // Won't type-check
```

**Comment by danog** [comment]
Bumping this, `psalm-if-this-is` is such a useful assertion, exactly what I need right now to solve an issue with wide template types, without splitting logic or suppressing issues :)

**Comment by olleharstedt** [comment]
Well, it breaks on aliasing, so it's not as tight as you might expect. There's another thread for @psalm-no-alias annotation, which might be a (much?) more complex issue.

**Comment by olleharstedt** [comment]
@danog You might want to have a look at the `@psalm-this-out` examples above, too.

**Comment by danog** [comment]
I just added a template-aware @psalm-this-out, check it out :D

**Comment by olleharstedt** [comment]
Yep, saw. :) Very nice, will close this issue.
